protocol/rpcprovider/rewardserver/reward_server.go

```diff
@@ -272,23 +354,136 @@ func (rws *RewardServer) PaymentHandler(payment *PaymentRequest) {
+
+		err = rws.rewardDB.DeleteClaimedRewards(payment.PaymentEpoch, payment.Client.String(), payment.UniqueIdentifier, payment.ConsumerRewardsKey)
+		if err != nil {
+			utils.LavaFormatWarning("failed deleting claimed rewards", err)
+		} else {
+			utils.LavaFormatDebug("deleted claimed rewards successfully")
 		}
 	}
 }
 
-func NewRewardServer(rewardsTxSender RewardsTxSender, providerMetrics *metrics.ProviderMetricsManager) *RewardServer {
-	//
+func NewRewardServer(rewardsTxSender RewardsTxSender, providerMetrics *metrics.ProviderMetricsManager, rewardDB *RewardDB, rewardStoragePath string, rewardsSnapshotThreshold uint, rewardsSnapshotTimeoutSec uint) *RewardServer {
 	rws := &RewardServer{totalCUServiced: 0, totalCUPaid: 0}
 	rws.serverID = uint64(rand.Int63())
 	rws.rewardsTxSender = rewardsTxSender
 	rws.expectedPayments = []PaymentRequest{}
-	// TODO: load this from persistency
-	rws.rewards = map[uint64]*EpochRewards{}
 	rws.providerMetrics = providerMetrics
+	rws.rewards = map[uint64]*EpochRewards{}
+	rws.rewardDB = rewardDB
+	rws.rewardStoragePath = rewardStoragePath
+	rws.rewardsSnapshotThreshold = uint64(rewardsSnapshotThreshold)
+	rws.rewardsSnapshotTimeoutDuration = time.Duration(rewardsSnapshotTimeoutSec * uint(time.Second))
+	rws.rewardsSnapshotTimer = timer.NewTimer(rws.rewardsSnapshotTimeoutDuration)
+	rws.rewardsSnapshotThresholdCh = make(chan struct{})
+
+	go rws.saveRewardsSnapshotToDBJob()
 	return rws
 }
 
+func (rws *RewardServer) saveRewardsSnapshotToDBJob() {
+	for {
+		select {
+		case <-rws.rewardsSnapshotTimer.C:
+			rws.resetSnapshotTimerAndSaveRewardsSnapshotToDBAndResetTimer()
+		case <-rws.rewardsSnapshotThresholdCh:
+			rws.resetSnapshotTimerAndSaveRewardsSnapshotToDBAndResetTimer()
+		}
+	}
+}
+
+func (rws *RewardServer) resetSnapshotTimerAndSaveRewardsSnapshotToDBAndResetTimer() {
+	// We lock without defer because the DB is already locking itself
+	rws.lock.RLock()
+	rws.rewardsSnapshotTimer.Reset(rws.rewardsSnapshotTimeoutDuration)
+
+	rewardEntities := []*RewardEntity{}
+	for epoch, epochRewards := range rws.rewards {
+		for consumerRewardKey, consumerRewards := range epochRewards.consumerRewards {
+			for sessionId, proof := range consumerRewards.proofs {
+				rewardEntity := &RewardEntity{
+					Epoch:        epoch,
+					ConsumerAddr: consumerRewards.consumer,
+					ConsumerKey:  consumerRewardKey,
+					SessionId:    sessionId,
+					Proof:        proof,
+				}
+				rewardEntities = append(rewardEntities, rewardEntity)
+			}
+		}
+	}
+	rws.lock.RUnlock()
+	if len(rewardEntities) == 0 {
+		return
+	}
+
+	utils.LavaFormatDebug("saving rewards snapshot to the DB", utils.Attribute{Key: "proofs", Value: len(rewardEntities)})
+
+	var err error
+	for i := 0; i < MaxDBSaveRetries; i++ {
+		err = rws.rewardDB.BatchSave(rewardEntities)
+		if err == nil {
+			return
+		}
+
+		utils.LavaFormatWarning("failed saving proofs snapshot to rewardDB. Retrying...", err,
```

Comment:
lets change it to INFO level instead of warning.

Reply:
What about the error?